### ISSUE TITLE
How to correctly write modules

### ISSUE THREAD

**Issue by twellspring** [saltstack/salt#42853]
### Description of Issue/Question
I am looking to extend the boto_elbv2 state using some already existing functions in the boto_elbv2 module.   While doing so I am getting errors in the module that do not get passed to the state results well.  I see the same problems with other states so it appears that this might be a common issue.

What is the correct way to write a module so that it will return the appropriate data when called directly or when called from a state?  Lots of modules that are using things like  `return (False, 'some error message')`  and the corresponding state does not pass through the error message to the state results.  I would like to write a well formatted state and update the corresponding module so that a state call returns a reasonable error message.

Is there any documentation on writing modules ( I could only find documentation on writing a state)?  Is there a well written state/module pair you can recommend that I can use a reference?

### Setup
Here is an example that seems pretty typical:

```
cron_delete_uploads:
    cron.present:
      - identifier: cron_delete_uploads
      - name:  find /tmp/uploads -type f -mmin +10
      - user: twells
      - minute: '*/5'
```

Trying to do a cron.present state when cron is not installed

```
          ID: cron_delete_uploads
    Function: cron.present
        Name: find /tmp/uploads -type f -mmin +10 -delete
      Result: False
     Comment: An exception occurred in this state: Traceback (most recent call last):
                File "/Users/twells/external_repos/salt/salt/state.py", line 1845, in call
                  **cdata['kwargs'])
                File "/Users/twells/external_repos/salt/salt/loader.py", line 1801, in wrapper
                  return f(*args, **kwargs)
                File "/Users/twells/external_repos/salt/salt/states/cron.py", line 332, in present
                  data = __salt__['cron.set_job'](user=user,
                File "/Users/twells/external_repos/salt/salt/loader.py", line 1120, in __getitem__
                  func = super(LazyLoader, self).__getitem__(item)
                File "/Users/twells/external_repos/salt/salt/utils/lazy.py", line 101, in __getitem__
                  raise KeyError(key)
              KeyError: 'cron.set_job'
     Started: 10:33:46.281907
    Duration: 15625.758 ms
     Changes:
```

I would expect to see `Cannot load cron module: crontab command not found` per the response in the cron.py module

```
def __virtual__():
    if salt.utils.which('crontab'):
        return (False, 'Cannot load cron module: crontab command not found')
```

### Steps to Reproduce Issue

Logs for above 
```
2017-08-10 11:17:47,731 [salt.utils.lazy                                                   ][DEBUG   ][21219] Could not LazyLoad cron.set_job: 'cron' __virtual__ returned False: Cannot load cron module: crontab command not found
2017-08-10 11:17:47,732 [salt.state                                                        ][ERROR   ][21219] An exception occurred in this state: Traceback (most recent call last):
  File "/Users/twells/external_repos/salt/salt/state.py", line 1845, in call
    **cdata['kwargs'])
  File "/Users/twells/external_repos/salt/salt/loader.py", line 1801, in wrapper
    return f(*args, **kwargs)
  File "/Users/twells/external_repos/salt/salt/states/cron.py", line 332, in present
    data = __salt__['cron.set_job'](user=user,
  File "/Users/twells/external_repos/salt/salt/loader.py", line 1120, in __getitem__
    func = super(LazyLoader, self).__getitem__(item)
  File "/Users/twells/external_repos/salt/salt/utils/lazy.py", line 101, in __getitem__
    raise KeyError(key)
KeyError: 'cron.set_job'

2017-08-10 11:17:47,732 [salt.state                                                        ][INFO    ][21219] Completed state [find /tmp -type f -mmin +10] at time 11:17:47.732635 duration_in_ms=14968.335
2017-08-10 11:17:47,733 [salt.state                                                        ][DEBUG   ][21219] File /Users/twells/python/salt/var/cache/salt/minion/accumulator/4414202512 does not exist, no need to cleanup.
2017-08-10 11:17:47,738 [salt.minion                                                       ][DEBUG   ][21219] Minion return retry timer set to 6 seconds (randomized)
2017-08-10 11:17:47,739 [salt.minion                                                       ][INFO    ][21219] Returning information for job: 20170810111731932378
2017-08-10 11:17:47,739 [salt.transport.zeromq                                             ][DEBUG   ][21219] Initializing new AsyncZeroMQReqChannel for ('/Users/twells/python/salt/etc/salt/pki/minion', 'Todd_Mac', 'tcp://127.0.0.1:4506', 'aes')
2017-08-10 11:17:47,740 [salt.crypt                                                        ][DEBUG   ][21219] Initializing new AsyncAuth for ('/Users/twells/python/salt/etc/salt/pki/minion', 'Todd_Mac', 'tcp://127.0.0.1:4506')
2017-08-10 11:17:47,752 [salt.minion                                                       ][DEBUG   ][21219] minion return: {'fun_args': ['test'], 'jid': '20170810111731932378', 'return': {'cron_|-cron_delete_uploads_|-find /tmp -type f -mmin +10_|-present': {'comment': 'An exception occurred in this state: Traceback (most recent call last):\n  File "/Users/twells/external_repos/salt/salt/state.py", line 1845, in call\n    **cdata[\'kwargs\'])\n  File "/Users/twells/external_repos/salt/salt/loader.py", line 1801, in wrapper\n    return f(*args, **kwargs)\n  File "/Users/twells/external_repos/salt/salt/states/cron.py", line 332, in present\n    data = __salt__[\'cron.set_job\'](user=user,\n  File "/Users/twells/external_repos/salt/salt/loader.py", line 1120, in __getitem__\n    func = super(LazyLoader, self).__getitem__(item)\n  File "/Users/twells/external_repos/salt/salt/utils/lazy.py", line 101, in __getitem__\n    raise KeyError(key)\nKeyError: \'cron.set_job\'\n', 'name': 'find /tmp -type f -mmin +10', 'start_time': '11:17:32.764300', 'result': False, 'duration': 14968.335, '__run_num__': 0, '__sls__': u'test', 'changes': {}, '__id__': 'cron_delete_uploads'}}, 'retcode': 2, 'success': True, 'fun': 'state.apply'}
```

### Versions Report

```
           Salt: 2017.7.0-415-gc000a2e

Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 2.6.1
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.9.6
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.8
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.10 (default, Oct 23 2015, 19:19:21)
   python-gnupg: Not Installed
         PyYAML: 3.12
          PyZMQ: 16.0.2
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.5.1
            ZMQ: 4.1.6

System Versions:
           dist:
         locale: UTF-8
        machine: x86_64
        release: 15.6.0
         system: Darwin
        version: 10.11.6 x86_64
```

**Comment by Ch3LL** [comment]
You will need to include the message in the state module. If you do a grep for `return (False` you will be able to see all the modules currently setup this way:

```
➜  states git:(0931281ebd) grep -ril "return (False" *
blockdev.py
boto_cfn.py
boto_iam.py
docker_container.py
docker_image.py
docker_network.py
docker.py
docker_volume.py
file.py
firewalld.py
heat.py
locale.py
netacl.py
netyang.py
postgres_cluster.py
postgres_database.py
postgres_extension.py
postgres_group.py
postgres_initdb.py
postgres_language.py
postgres_privileges.py
postgres_schema.py
postgres_tablespace.py
postgres_user.py
reg.py
service.py
x509.py
```

Also [here](https://docs.saltstack.com/en/latest/ref/modules/) is some documentation on writing modules.Particularly right [here](https://docs.saltstack.com/en/latest/ref/modules/#returning-error-information-from-virtual) it explains the need to add this error message in both execution and state module

**Comment by terminalmage** [comment]
Another good way of doing this is to raise a `CommandExecutionError`

```python
from salt.exceptions import CommandExecutionError

raise CommandExecutionError(
    'Some message here',
    info={'errors': errors, 'changes': changes}
)
```

This is just an example, you don't need to pass `errors` or `changes` in the `info` param. The data structure passed as `info` will be assigned to the raised exception as its `info` attribute. So then, in states, you can do this:

```python
from salt.exceptions import CommandExecutionError

try:
    result = __salt__['mymodule.myfunction'](some_arg, another_arg)
except CommandExecutionError as exc:
    if exc.info:
        do some stuff here
```

Check `salt/states/pkg.py` for examples of this approach being taken. Using this approach lets us report on partial changes even when error conditions were reached in the process of running the state.

**Comment by terminalmage** [comment]
Actually, it looks like like you're more having issues because the supporting module is not present. That message that gets returned in the cron module's `__virtual__` func  is simply logged at the `trace` loglevel, the state doesn't have access to those returns because they happen in the loader.

The way we handle this typically is to look for a needed function in the state module's `__virtual__` function, and conditionally load the state module based on whether that function is available. I noticed that this was never done for the cron state, so I added this in https://github.com/saltstack/salt/pull/42877. This will result in the cron state module not being available. It won't tell you exactly what is wrong though, for this you would need to run the minion with `log_level_logfile: trace` to see the error.
